stm32: u5: allow selection of VOS via DT

mathieuchopstm opened the pull request:
Allow selecting a voltage scale using Device Tree on STM32U5 series. This is notably useful when a high system clock frequency is not required, but the application still wants to use the USB IPs which are only available in certain ranges.

Fixes #98641.

mathieuchopstm committed:
dts: bindings: pwr: add STM32U5 power controller binding

Add a binding for the power controller of STM32U5 series.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: clock_control: stm32u5: add configurable voltage scale support

Add support for the new `voltage-scale` property on the STM32U5-specific
Power Controller binding.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
release-notes: 4.4: add new `voltage-scale` property for STM32U5 PWRC

Add new voltage scale selection property for STM32U5 to 4.4 release notes.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

petrosyan-van commented:
> @petrosyan-van could you please give some feedback on this? 🙂

Thanks for this PR! I will try it out in upcoming days and will let you know :)

mathieuchopstm commented on dts/bindings/power/st,stm32u5-pwr.yaml:
```diff
@@ -0,0 +1,58 @@
+      +---------+----------------+-----------------------------------+
+      |    3    |     55  MHz    | [1]                               |
+      +---------+----------------+-----------------------------------+
+      |    4    |     25  MHz    | [1][2]; lowest power consumption  |
```
The code in UDC driver is overzealous - the table is correct and based on STM32U5 series Reference Manual:
<img width="936" height="944" alt="image" src="https://github.com/user-attachments/assets/079d8b09-b0a4-40c3-a87b-162209188960" />

The UDC driver should definitely be fixed, but this is a separate issue outside PR scope.

erwango reviewed (changes requested):
Otherwise LTGM

petrosyan-van commented:
I tried with `voltage-scale = <2>;` and the USB stack now works, thanks! :)
Is there a possibility to change it dynamically? I mean, ideally, I would set the `voltage-scale` to `2` (or `3` after UDC driver fix), when USB is connected, but would go to `4` when it is disconnected to save some power. Cause my plan is to enable/disable USB stack on plug/unplug

mathieuchopstm committed:
dts: arm: st: stm32: u5: add U5-specific compatible on Power Controller

Add the "st,stm32u5-pwr" compatible to the Power Controller node in the
DTSI for STM32U5 series.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>

mathieuchopstm committed:
drivers: clock_control: stm32u5: fix frequency range check for scale

The range check should be inclusive: scale 4 allows *up to* 25 MHz, scale 3
*up to* 55 MHz, etc.

Signed-off-by: Mathieu Choplain <mathieu.choplain-ext@st.com>
